lib/gst/player/gstplayer-media-info.c

```diff
@@ -0,0 +1,649 @@
+gst_player_subtitle_info_copy (GstPlayerSubtitleInfo *ref)
+{
+  GstPlayerSubtitleInfo *ret;
+
+  ret = gst_player_subtitle_info_new ();
+  if (ref->language)
+    ret->language = g_strdup (ref->language);
+
+  return (GstPlayerStreamInfo*) ret;
+}
+
+GstPlayerStreamInfo*
+gst_player_stream_info_copy (GstPlayerStreamInfo *ref)
+{
+  GstPlayerStreamInfo *info = NULL;
+
+  if (GST_IS_PLAYER_VIDEO_INFO(ref))
+    info = gst_player_video_info_copy ((GstPlayerVideoInfo*)ref);
+
+  if (GST_IS_PLAYER_AUDIO_INFO(ref))
+    info = gst_player_audio_info_copy ((GstPlayerAudioInfo*)ref);
+
+  if (GST_IS_PLAYER_SUBTITLE_INFO(ref))
+    info = gst_player_subtitle_info_copy ((GstPlayerSubtitleInfo*)ref);
+
+  info->stream_index = ref->stream_index;
+  if (ref->tags)
+    info->tags = gst_tag_list_copy (ref->tags);
+  if (ref->caps)
+    info->caps = gst_caps_copy (ref->caps);
+
+  return info;
+}
+
+GstPlayerMediaInfo*
+gst_player_media_info_copy (GstPlayerMediaInfo* ref)
+{
+  GList *l;
+  GstPlayerMediaInfo *info;
+
+  if (!ref)
+    return NULL;
+
+  info = gst_player_media_info_new (ref->uri);
+
+  for (l = ref->stream_list; l != NULL; l = l->next)
+    info->stream_list = g_list_append (info->stream_list,
+                  gst_player_stream_info_copy((GstPlayerStreamInfo*)l->data));
+
+  return info;
+}
+
+GstPlayerStreamInfo*
+gst_player_stream_info_new (gint stream_index, GType type)
+{
+  GstPlayerStreamInfo *info = NULL;
+
+  if (type == GST_TYPE_PLAYER_AUDIO_INFO)
+    info = (GstPlayerStreamInfo*) gst_player_audio_info_new ();
+
+  if (type == GST_TYPE_PLAYER_VIDEO_INFO)
+    info = (GstPlayerStreamInfo*) gst_player_video_info_new ();
+
+  if (type == GST_TYPE_PLAYER_SUBTITLE_INFO)
+    info = (GstPlayerStreamInfo*) gst_player_subtitle_info_new ();
+
+  info->stream_index = stream_index;
+
+  return info;
+}
+
+GstPlayerMediaInfo*
+gst_player_media_info_new (const gchar *uri)
+{
+  GstPlayerMediaInfo *info;
+
+  info = g_object_new (GST_TYPE_PLAYER_MEDIA_INFO, NULL);
+  if (uri)
+    info->uri = g_strdup (uri);
+
+  return info;
+}
+
+/**
+ * gst_player_media_info_get_uri:
+ * @info: a #GstPlayerMediaInfo
+ *
+ * Returns: the URI associated with #GstPlayerMediaInfo.
+ */
+const gchar *
+gst_player_media_info_get_uri (const GstPlayerMediaInfo *info)
+{
+  g_return_val_if_fail (GST_IS_PLAYER_MEDIA_INFO (info), NULL);
+
+  return info->uri;
+}
+
+/**
+ * gst_player_media_info_get_stream_list:
+ * @info: a #GstPlayerMediaInfo
+ *
+ * Returns: (transfer none) (element-type GstPlayerStreamInfo): A #GList of
+ * matching #GstPlayerStreamInfo. The caller should free it with
+ * gst_player_stream_info_list_free().
+ */
+GList *
+gst_player_media_info_get_stream_list (const GstPlayerMediaInfo *info)
+{
+  g_return_val_if_fail (GST_IS_PLAYER_MEDIA_INFO (info), NULL);
+
+  return info->stream_list;
+}
+
+/**
+ * gst_player_get_video_streams:
+ * @info: a #GstPlayerMediaInfo
+ *
+ * Returns: (transfer full) (element-type GstPlayerVideoInfo): A #GList of
```

Comment:
sure sound like good idea. I will create GList *_cachelist in GstPlayerMediaInfo which will point to audio, video and subtitle list and then we can simply return that *_cachelist from here.